ml-agents/mlagents/trainers/bc/trainer.py

```diff
@@ -42,67 +41,45 @@ def __init__(self, brain, trainer_parameters, training, load, seed, run_id):
-        self.evaluation_buffer = ProcessingBuffer()
-
-    def add_experiences(
-        self,
-        curr_info: BrainInfo,
-        next_info: BrainInfo,
-        take_action_outputs: ActionInfoOutputs,
-    ) -> None:
-        """
-        Adds experiences to each agent's experience history.
-        :param curr_info: Current BrainInfo
-        :param next_info: Next BrainInfo
-        :param take_action_outputs: The outputs of the take action method.
-        """
 
-        # Used to collect information about student performance.
-        for agent_id in curr_info.agents:
-            self.evaluation_buffer[agent_id].last_brain_info = curr_info
-
-        for agent_id in next_info.agents:
-            stored_next_info = self.evaluation_buffer[agent_id].last_brain_info
-            if stored_next_info is None:
-                continue
-            else:
-                next_idx = next_info.agents.index(agent_id)
-                if agent_id not in self.cumulative_rewards:
-                    self.cumulative_rewards[agent_id] = 0
-                self.cumulative_rewards[agent_id] += next_info.rewards[next_idx]
-                if not next_info.local_done[next_idx]:
-                    if agent_id not in self.episode_steps:
-                        self.episode_steps[agent_id] = 0
-                    self.episode_steps[agent_id] += 1
-
-    def process_experiences(
-        self, current_info: BrainInfo, next_info: BrainInfo
-    ) -> None:
+    def process_trajectory(self, trajectory: Trajectory) -> None:
         """
-        Checks agent histories for processing condition, and processes them as necessary.
+        Takes a trajectory and processes it, putting it into the update buffer.
         Processing involves calculating value and advantage targets for model updating step.
-        :param current_info: Current BrainInfo
-        :param next_info: Next BrainInfo
         """
-        for l in range(len(next_info.agents)):
-            if next_info.local_done[l]:
-                agent_id = next_info.agents[l]
-                self.stats["Environment/Cumulative Reward"].append(
-                    self.cumulative_rewards.get(agent_id, 0)
-                )
-                self.stats["Environment/Episode Length"].append(
-                    self.episode_steps.get(agent_id, 0)
-                )
-                self.reward_buffer.appendleft(self.cumulative_rewards.get(agent_id, 0))
-                self.cumulative_rewards[agent_id] = 0
-                self.episode_steps[agent_id] = 0
+        agent_id = trajectory.agent_id  # All the experiences should have the same ID
+        agent_buffer_trajectory = trajectory_to_agentbuffer(trajectory)
+
+        # Evaluate all reward functions
+        self.cumulative_rewards[agent_id] += np.sum(
+            agent_buffer_trajectory["environment_rewards"]
+        )
+
+        # Increment episode steps
+        if agent_id not in self.episode_steps:
```

Comment:
defaultdict for episode_steps too?

Reply:
Actually, collections.Counter might be better suited.

Reply:
I'm using defaultdict in the stats reporter [PR](https://github.com/Unity-Technologies/ml-agents/pull/3076/files#diff-8db9ccc978b4e64b2ea57498841f97afL37). Will look into Counter for this application